edx/edx-enterprise
# ENT-182: Defer EnterpriseCustomerUser creation until after consent is provided

**Description:** This pull request refactors the `post_account_consent` method to ensure that new users, entering the system via SSO, are not linked to an EnterpriseCustomer unless they provide the relevant level of data sharing consent.

**JIRA:** [ENT-182](https://openedx.atlassian.net/browse/ENT-182)

**Dependencies:** None

**Merge deadline:** 27 February 2017

**Installation instructions:** Trivial installation; standard features. SSO provider setup required.

**Sandbox link**: https://ent182.sandbox.edx.org/; use staff/edx user for admin access

**Testing instructions:**

1. Set up a SAML integration with TestShib
2. Set up an EnterpriseCustomer linked to that SAML integration
3. Register a user account using standard methods, without using the SSO integration.
4. In a new incognito tab, browse to the instance main page and click Login.
5. Click on the TestShib SSO button; proceed through the login flow.
6. When redirected back to the instance, you'll be presented with an option to log in with your username and password. Log in with the username and password of the account you created earlier.
7. Because the SSO provider is linked to your EnterpriseCustomer, you'll be prompted for data sharing consent. Do not check the consent box, and click Submit.
8. You will be redirected back to the login page, with your browsing session flushed.
9. In a separate window, viewing the Manage Learners admin page of your EnterpriseCustomer, observe that the user you just logged in with is _not_ linked to the EnterpriseCustomer.
10. Observe that if you attempt to log in with the SSO provider, you are prompted for consent. Provide it.
11. Observe that only after providing consent in the previous step, the user you created is added to the list of linked learners in the Manage Learners admin page.

**Reviewers:**

**Merge checklist:**

**Post merge:**
      finished.
